### PR TITLE
HPCC-21167 Report in logs if transparent_hugepage/enabled is set to always

### PR DESCRIPTION
**Pull request by ghalliday**
* Also ensure pages are returned to the system in multiples of huge page size
  if transparent_hugepage/enabled set to always

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln please review

File: roxie/roxiemem/roxiemem.cpp [hunk 1]
```diff
@@ -292,8 +292,19 @@ static void initializeHeap(bool allowHugePages, bool allowTransparentHugePages,
     if (!heapBase)
     {
         const memsize_t hugePageSize = getHugePageSize();
-        bool useTransparentHugePages = allowTransparentHugePages && areTransparentHugePagesEnabled();
+        HugePageMode mode = queryTransparentHugePagesMode();
+        bool hasTransparenHugePages = (mode == HugePageMode::Always) || (mode == HugePageMode::Madvise);
+        bool useTransparentHugePages = allowTransparentHugePages && hasTransparenHugePages;
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply roxie/roxiemem/roxiemem.cpp]: no, but I didn't want to call a function to read the value more than once.  I could really delete areTransparentHugePagesEnabled()

[the remark this answers is not in the document]
Comment (mckellyln) [reply roxie/roxiemem/roxiemem.cpp]: Sure, delete it.

[the remark this answers is not in the document]
Comment (ghalliday) [reply roxie/roxiemem/roxiemem.cpp]: Changed it.

File: roxie/roxiemem/roxiemem.cpp [hunk 2]
```diff
@@ -292,8 +292,19 @@ static void initializeHeap(bool allowHugePages, bool allowTransparentHugePages,
...
         memsize_t heapAlignment = useTransparentHugePages ? hugePageSize : HEAP_ALIGNMENT_SIZE;
+        if (mode == HugePageMode::Always)
+        {
+            //Always return memory in multiples of the huge page size - even if it is not being used
+            heapAlignment = hugePageSize;
+            ERRLOG("WARNING: The OS is configured to always use transparent huge pages.  This may cause unexplained pauses "
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply roxie/roxiemem/roxiemem.cpp]: I think it should now be reported...  please check

**Comment by ghalliday** [comment]
@mckellyln please check modified code.

File: roxie/roxiemem/roxiemem.cpp [hunk 3]
```diff
@@ -292,8 +292,19 @@ static void initializeHeap(bool allowHugePages, bool allowTransparentHugePages,
...
-        bool useTransparentHugePages = allowTransparentHugePages && areTransparentHugePagesEnabled();
+        HugePageMode mode = queryTransparentHugePagesMode();
+        bool hasTransparenHugePages = areTransparentHugePagesEnabled(mode);
+        bool useTransparentHugePages = allowTransparentHugePages && hasTransparenHugePages;
         memsize_t heapAlignment = useTransparentHugePages ? hugePageSize : HEAP_ALIGNMENT_SIZE;
+        if (mode == HugePageMode::Always)
+        {
+            //Always return memory in multiples of the huge page size - even if it is not being used
+            heapAlignment = hugePageSize;
+            OERRLOG("WARNING: The OS is configured to always use transparent huge pages.  This may cause unexplained pauses "
```
Comment (mckellyln) [inline roxie/roxiemem/roxiemem.cpp]: Three separate strings, is this ok?

Comment (mckellyln) [reply roxie/roxiemem/roxiemem.cpp]: ok, 3 strings format is ok - but I dont see it in my syslog msgs.  Let me debug.

Comment (mckellyln) [reply roxie/roxiemem/roxiemem.cpp]: ok, so in Thorslave for example, is seems UseSysLogForOperatorMessages() is not called and so OERRLOG() does not show up in syslog.  Should we add
```
        if (globals->getPropBool("@enableSysLog",true))
            UseSysLogForOperatorMessages();
```
to Thorslave in thslavemain.cpp in or after startSlaveLog() ?

Comment (ghalliday) [reply roxie/roxiemem/roxiemem.cpp]: they get concatenated

Comment (jakesmith) [reply roxie/roxiemem/roxiemem.cpp]: hm, yes slaves should have on too (I thought did).

Yes, I think should add, but separate issue.

**Review by mckellyln — APPROVED** [review]
Approved.
But we do need to add an issue for Thorslave to use syslog, if enabled.

**Comment by mckellyln** [comment]
@ghalliday you mention
```
Also ensure pages are returned to the system in multiples of huge page size
if transparent_hugepage/enabled set to always
```
Is there new / additional code for this ?

File: roxie/roxiemem/roxiemem.cpp [hunk 4]
```diff
@@ -292,8 +292,19 @@ static void initializeHeap(bool allowHugePages, bool allowTransparentHugePages,
...
         const memsize_t hugePageSize = getHugePageSize();
-        bool useTransparentHugePages = allowTransparentHugePages && areTransparentHugePagesEnabled();
+        HugePageMode mode = queryTransparentHugePagesMode();
+        bool hasTransparenHugePages = areTransparentHugePagesEnabled(mode);
+        bool useTransparentHugePages = allowTransparentHugePages && hasTransparenHugePages;
         memsize_t heapAlignment = useTransparentHugePages ? hugePageSize : HEAP_ALIGNMENT_SIZE;
+        if (mode == HugePageMode::Always)
+        {
+            //Always return memory in multiples of the huge page size - even if it is not being used
+            heapAlignment = hugePageSize;
```
Comment (mckellyln) [inline roxie/roxiemem/roxiemem.cpp]: yes, probably a very good idea

**Comment by ghalliday** [comment]
@richardkchapman I think this is now ready to merge

**Comment by mckellyln** [comment]
https://track.hpccsystems.com/browse/HPCC-21217 created

**Pull request merged**